File: nicoleWojewoda/assignments/twopointers/lc977/lc977.js
// Problem Statement #
// Given a sorted array, create a new array containing squares of all the numbers of the input array in the sorted order.

/*
input: a sorted array
output: a sorted array containing the square of each element

brute-force:
steps:
  init output array
  loop through the given array
    square each element, add it to the output array
   
  sort the output array
  return the output array

better:
steps:
  init output array, left pointer, right pointer, position pointer
  start left pointer at 0 and right pointer at arr.length-1
  start pos pointer at arr.length-1
  loop while r >= l
    if r^2 is >= l^2,
      add r^2 to pos's index in the result array
      decrement right
    else, 
      add l^2 to pos's index in the result array
      increment left

    decrement pos
  
  return the result array
*/

function make_squares(arr) {
  if (!arr) return [];

  let result = Array(arr.length);
  let left = 0;
  let right = arr.length - 1;
  let pos = arr.length - 1;

  while (right >= left) {
    if (arr[right] * arr[right] >= arr[left] * arr[left]) {
      result[pos] = arr[right] * arr[right];
      right--;
    } else {
      result[pos] = arr[left] * arr[left];
      left++;
    }

    pos--;
  }

  return result;
}

console.log(`Squares: ${make_squares([-2, -1, 0, 2, 3])}`); // [0, 1, 4, 4, 9]
console.log(`Squares: ${make_squares([-3, -1, 0, 1, 2])}`); // [0, 1, 1, 4, 9]

// Solution
// -----
// const n = arr.length;
// squares = Array(n).fill(0);
// let highestSquareIdx = n - 1;
// let left = 0,
//   right = n - 1;
// while (left <= right) {
//   let leftSquare = arr[left] * arr[left],
//     rightSquare = arr[right] * arr[right];
//   if (leftSquare > rightSquare) {
//     squares[highestSquareIdx] = leftSquare;
//     left += 1;
//   } else {
//     squares[highestSquareIdx] = rightSquare;
//     right -= 1;
//   }
//   highestSquareIdx -= 1;
// }

// return squares;

// -----

// Time complexity #
// The above algorithm’s time complexity will be O(N) as we are iterating the input array only once.

// Space complexity #
// The above algorithm’s space complexity will also be O(N); this space will be used for the output array.

File: xiaqianZhang/assignments/twopointers/lc977/lc977.py
# Problem Statement #
# Given a sorted array, create a new array containing squares of all the numbers of the input array in the sorted order.

'''
Input: array of integers that includes negative values
Output: return the new sorted array that contains squares of each integers of the original array

edge case: if the array is empty, return []

Only iterating the array for once so the running time should be O(N):
everynumber that is less than 0 should be times by -1 [2, 1, 0, 2, 3]

startIndex = 0
lastIndex = len(arr) - 1
initliaze a result squared array = []
while startIndex < lastIndex: when they are not meet
  check if the squared of the startIndex > squared of the lastIndex: [-3, 2] -> [9, 4]
    so the original value at the lastIndex should be the largest value, then if the square of the startIndex is larger than, then this startIndex should be the largest
    squared array.insert(0, arr[startIndex]) decreasing order?
    increment the startIndex

  if the squared lastIndex is larger than the startIndex:
    then the lastIndex's value is still the largest, append it into the squared array
    decrement the lastIndex
  
  if they are the same, append startIndex, then increment the startIndex

'''
def make_squares(arr):
  squares = []
  startIndex = 0
  lastIndex = len(arr) -1
  if len(arr) == 0:
    return []
  
  while startIndex <= lastIndex:
    
    if arr[startIndex] ** 2 > arr[lastIndex] ** 2:
      squares.insert(0, arr[startIndex] ** 2)
      startIndex+=1
    else:
      squares.insert(0, arr[lastIndex] ** 2)
      lastIndex-=1
  return squares


def main():

  print("Squares: " + str(make_squares([-2, -1, 0, 2, 3])))
  print("Squares: " + str(make_squares([-3, -1, 0, 1, 2])))
  #[, 4, 9]
  print("Squares: " + str(make_squares([])))
  print("Squares: " + str(make_squares([-3, -1, 0, 0, 1, 2])))


main()








# Solution
# -----
#   n = len(arr)
#   squares = [0 for x in range(n)]
#   highestSquareIdx = n - 1
#   left, right = 0, n - 1
#   while left <= right:
#     leftSquare = arr[left] * arr[left]
#     rightSquare = arr[right] * arr[right]
#     if leftSquare > rightSquare:
#       squares[highestSquareIdx] = leftSquare
#       left += 1
#     else:
#       squares[highestSquareIdx] = rightSquare
#       right -= 1
#     highestSquareIdx -= 1

#   return squares

#   -----

#   Time complexity #
# The above algorithm’s time complexity will be O(N) as we are iterating the input array only once.

# Space complexity #
# The above algorithm’s space complexity will also be O(N); this space will be used for the output array.
